huaweicloud/services/nat/data_source_huaweicloud_nat_private_transit_ips.go

```diff
@@ -0,0 +1,259 @@
+	return &sc
+}
+
+func dataSourcePrivateTransitIpsRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+
+	// listTransitIps: Query the transit IP list
+	var (
+		listTransitIpsHttpUrl = "v3/{project_id}/private-nat/transit-ips"
+		listTransitIpsProduct = "nat"
+	)
+	listTransitIpsClient, err := cfg.NewServiceClient(listTransitIpsProduct, region)
+	if err != nil {
+		return diag.Errorf("error creating NAT client: %s", err)
+	}
+
+	listTransitIpsPath := listTransitIpsClient.Endpoint + listTransitIpsHttpUrl
+	listTransitIpsPath = strings.ReplaceAll(listTransitIpsPath, "{project_id}", listTransitIpsClient.ProjectID)
+
+	listTransitIpsQueryParams := buildListTransitIpsQueryParams(d, cfg)
+	listTransitIpsPath += listTransitIpsQueryParams
+
+	listTransitIpsResp, err := pagination.ListAllItems(
+		listTransitIpsClient,
+		"marker",
+		listTransitIpsPath,
+		&pagination.QueryOpts{MarkerField: ""})
+
+	if err != nil {
+		return common.CheckDeletedDiag(d, err, "error retrieving transit IPs")
+	}
+
+	listTransitIpsRespJson, err := json.Marshal(listTransitIpsResp)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	var listTransitIpsRespBody interface{}
+	err = json.Unmarshal(listTransitIpsRespJson, &listTransitIpsRespBody)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	uuid, err := uuid.GenerateUUID()
+	if err != nil {
+		return diag.Errorf("unable to generate ID: %s", err)
+	}
+	d.SetId(uuid)
+
+	curJson := utils.PathSearch("transit_ips", listTransitIpsRespBody, make([]interface{}, 0))
+	curArray := curJson.([]interface{})
+
+	var mErr *multierror.Error
+	mErr = multierror.Append(
+		mErr,
+		d.Set("region", region),
+		d.Set("transit_ips", flattenListTransitIpsResponseBody(filterListTransitIpsResponseBody(curArray, d))),
+	)
+
+	return diag.FromErr(mErr.ErrorOrNil())
+}
+
+func flattenListTransitIpsResponseBody(resp []interface{}) []interface{} {
+	if resp == nil {
+		return nil
+	}
+
+	rst := make([]interface{}, 0, len(resp))
+	for _, v := range resp {
+		rst = append(rst, map[string]interface{}{
+			"id":                    utils.PathSearch("id", v, nil),
+			"ip_address":            utils.PathSearch("ip_address", v, nil),
+			"subnet_id":             utils.PathSearch("virsubnet_id", v, nil),
+			"gateway_id":            utils.PathSearch("gateway_id", v, nil),
+			"tags":                  utils.FlattenTagsToMap(utils.PathSearch("tags", v, nil)),
+			"created_at":            utils.PathSearch("created_at", v, nil),
+			"updated_at":            utils.PathSearch("updated_at", v, nil),
+			"network_interface_id":  utils.PathSearch("network_interface_id", v, nil),
+			"enterprise_project_id": utils.PathSearch("enterprise_project_id", v, nil),
+		})
+	}
+	return rst
+}
+
+func filterListTransitIpsResponseBody(all []interface{}, d *schema.ResourceData) []interface{} {
+	rst := make([]interface{}, 0, len(all))
+	tagFilter := d.Get("tags").(map[string]interface{})
```

Comment:
In most situation, the `tags` param is not specified, so we can return directly when it is empty.
```suggestion
	tagFilter := d.Get("tags").(map[string]interface{})
	if len(tagFilter) == 0 {
		return all
	}
```

Reply:
done